src/codeblockediting.js

```diff
@@ -0,0 +1,409 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module code-block/codeblockediting
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+import ShiftEnter from '@ckeditor/ckeditor5-enter/src/shiftenter';
+import CodeBlockCommand from './codeblockcommand';
+import IndentCodeBlockCommand from './indentcodeblockcommand';
+import {
+	getLocalizedLanguageDefinitions,
+	getLeadingWhiteSpaces,
+	rawSnippetTextToModelDocumentFragment
+} from './utils';
+import {
+	modelToViewCodeBlockInsertion,
+	modelToDataViewSoftBreakInsertion,
+	dataViewToModelCodeBlockInsertion
+} from './converters';
+
+const DEFAULT_ELEMENT = 'paragraph';
+
+/**
+ * The editing part of the code block feature.
+ *
+ * Introduces the `'codeBlock'` command and the `'codeBlock'` model element.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class CodeBlockEditing extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'CodeBlockEditing';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ ShiftEnter ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		editor.config.define( 'codeBlock', {
+			languages: [
+				{ class: 'plaintext', label: 'Plain text' },
```

Comment:
It looks like both [highlight.js](https://highlightjs.org/usage/) and [prism](https://prismjs.com/#basic-usage) support `language-foo` syntax. Maybe to make people's lives easier, let's provide classes compatible with those highlighters by default? They wouldn't need to modify the config then.

Reply:
Google's code-prettify also follows this rule when it is set on `<code>` element: https://github.com/google/code-prettify/blob/master/src/prettify.js#L1630 quoting this: https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-code-element. So going with `language-foo` is a way to go :)